packages/web3-core-helpers/src/errors.js

```diff
@@ -43,5 +43,12 @@ module.exports = {
     },
     ConnectionTimeout: function (ms){
         return new Error('CONNECTION TIMEOUT: timeout of ' + ms + ' ms achived');
+    },
+    RevertInstructionError: function(reason, signature) {
+        var error = new Error('Your request got reverted with the following reason string: ' + reason);
```

Comment:
> Lots of people grep that and parse the receipt out of the string as well.

Because the new error message only appears if the ``handleRevert`` property is set to ``true`` do I think can we keep the error message as it is.  

> And maybe to handle .call you would have the function flow be

I thought to keep the logic to only return the receipt for the error listener and to not duplicate the return values with adding it to the ``Error`` object. If the mentioned above is the wished behavior do I propose because of consistency reasons to add the ``receipt`` to all ``Error`` objects.

Reply:
> Because the new error message only appears if the handleRevert property is set to true do I think can we keep the error message as it is.

What is the advantage of your design?  

>  add the receipt to all Error objects

Yes that seems reasonable to me. 

About serializing the receipt in the message - that decision was made a long time ago and people have adapted to it. They've written all kinds of logic to manage it. Suddenly deciding not to support that interface seems weird to me, idk.

Reply:
> What is the advantage of your design?

It doesn't break existing code and will only break it if you use the newly provided property which was, in the beginning, the reason to add it. 

> Yes that seems reasonable to me.

Great I will open a separate PR for this!

> About serializing the receipt in the message - that decision was made a long time ago and people have adapted to it. They've written all kinds of logic to manage it. Suddenly deciding not to support that interface seems weird to me, idk.

I've used another error message to only have one new ``Error`` which is used on ``eth_call`` and ``eth_sendTransaction``. If the serialized receipt is a must-have for those error messages can we also have two different ``Error``'s. This means I would use the old one (``Transaction has been reverted by the EVM:\n' + receiptJSON'``) for ``eth_sendTransaction`` calls and the new one for direct ``eth_call`` calls.

Reply:
This would probably allow us to remove the ``handleRevert`` options property at all because the ``Error`` is still the same and does only have additional properties if a revert message is existing.

Reply:
> I've used another error message to only have one new Error which is used on eth_call and eth_sendTransaction. If the serialized receipt is a must-have for those error messages can we also have two different Error's. 

Wait I don't understand that part. Keeping the error formats similar doesn't seem like it causes lots of problems. 

> This would probably allow us to remove the handleRevert options property at all because the Error is still the same and does only have additional properties if a revert message is existing.

I don't think this true because making additional calls that are already duplicated upstream for this lib potentially slows things down, so it should be optional. There's also completely different behavior for `.call` and potentially some security considerations for Solidity writers.

Reply:
I will keep the ``handleRevert`` property as it is and I've added the ``TransactionRevertInstructionError``. This error does have the same error message as we currently do have. With the difference of two additional added object properties called ``reason`` and ``signature``.

Reply:
I'm unresolving this to look at it again tomorrow, sorry.

Reply:
> some security considerations for Solidity writers.

I can't really see what security considerations are here because someone could also call the Ethereum blockchain with his own node and scripts and would get those details returned. There is nothing I think web3.js has to do in this direction. 

> Keeping the error formats similar doesn't seem like it causes lots of problems.

I've chosen a generic error message because the same error was used for ``eth_call`` and ``eth_sendTransaction``. I have created now, in addition, a second error especially for the transaction case which does have the same properties as the generic one but with the small difference of having the same error message as we have currently. (``'Transaction has been reverted by the EVM:\n' + receiptJSON``)

Reply:
I think it makes sense to be **as consistent as possible with the existing implementation**. The library has many existing users who want to see the reason. They'd also like to get it while making the smallest number changes to their own code. That means:

+ New error messages should include the reason. (If this isn't done, there will immediately be issues opened saying feature doesn't work).
+ New error messages should resemble the old message. (So existing handlers are not broken.)
+ New error message for `call` and `send` should resemble each other as much a possible. (So, similar error handlers can be written/used. DRY.) 

I think the above conditions are difficult to get perfect but atm the logic in this PR does not satisfy these requirements.

> I can't really see what security considerations are here because someone could also call the Ethereum....There is nothing I think web3.js has to do in this direction.

Ok, I'm not sure how easy it is to do the fake error I suggested below anyway. But perhaps we could get more opinions about the security implications. I think Web3 *should* worry a little about what it's like to use the library without understanding all the small details:

+ Most users assume that error messages are 'official' and believe them. 
+ Many engineers would be quite surprised if their users could modify the contract such that ordinary data was interpreted by the JS API as an error with a custom message the contract writers had not approved.

Reply:
All points above are implemented as described with the last commit. The error message is still the same I've only extended the object property list with ``signature`` and ``reason`` for the ``eth_sendTransaction`` call. The only difference and also only if ``handleRevert`` is set to true is that ``eth_call`` does throw an error if a revert instruction got detected.

 > Many engineers would be quite surprised if their users could modify the contract such that ordinary data was interpreted by the JS API as an error with a custom message the contract writers had not approved.

This is something we never really can pretend in JS. The only way to pretend a user to modify a contract object would be if web3.js would only run in a JavaScript runtime which does have SES and the related Realms object implemented. The other case you mentioned with spoofing of the requests and returning of wrong errors (malicious node) is also something we only could partially pretend with having of simultaneously running providers which are connected to different nodes and we compare the returned result of them. A normal middle man attack can be pretended completely by our users using SSL encrypted communication between the client and the connected node.

Reply:
> All points above are implemented as described with the last commit. The error message is still the same I've only extended the object property list with signature and reason for the eth_sendTransaction call.

Here are the points again:

+  New error messages should include the reason. (If this isn't done, there will immediately be issues opened saying feature doesn't work).
+ New error messages should resemble the old message. (So existing handlers are not broken.)
+ New error message for `call` and `send` should resemble each other as much a possible. (So, similar error handlers can be written/used. DRY.) 

He are the ways it's currently different from the points:
1. The TransactionRevertInstructionError does not include the reason in the error message string. 
2. (Only making this point so that when 1 is done, 2 is also satisfied.)
3. `call` message contains the reason in the error message string, `send` does not. 

Also the message for `call` is this:
```
Your request got reverted with the following reason string:
```

It might make sense to reserve the word 'revert' for state-changing transactions. Most `calls` will just be to read from a view fn. Writing the message so it captures that distinction could be helpful. I would change my original wording for the reason phrase to:
```
Request errored with Solidity reason: ' + reason
```

RE: MITM etc, ok fair enough although I think those issues affect all applications - this question is quite specific to expectations about how Solidity behaves that might not be clear to everyone. I'm not suggesting that the call handling be removed - only that this issue is discussed with others to get their opinion about it's severity and to think about how to disclose it to users (if necessary).

Reply:
> The TransactionRevertInstructionError does not include the reason in the error message string.

Yes, to not break the existing code. I would keep this as it is and we can inform the user about the interface in the release and our documentation. The proposed error message from you would be a really ugly solution which I will not implement.

> call message contains the reason in the error message string, send does not.

Yes, because it is a completely new error and does anyways throw where it wasn't throwing an error before. I think this is totally okay to have it like this. 

> Request errored with Solidity reason: ' + reason

This would exclude Vyper and other contract languages. The word ``reverted`` is also correct because the revert instruction got executed in the EVM.

Reply:
Good point about Vyper.  I also see what you mean about the opcode as well - there's a case there. 

> The proposed error message from you would be a really ugly solution which I will not implement.

Ok. This seems like the root of the disagreement. This discussion has been a waste of everyone's time.

Reply:
> This discussion has been a waste of everyone's time.

Yes, the learnings I take away from this is that also changing of an error message string does count as breaking change and that such changes in the "heart" of web3.js should be discussed in detail before the implementation does start. This will probably save time on both sides the proposer and reviewer of a PR. Thanks for your inputs and the review at all!